File: sdks/python/src/attest/_proto/codec.py
"""JSON-RPC 2.0 codec for Attest protocol communication."""

from __future__ import annotations

import json
from typing import Any

from attest._proto.types import ErrorData


class ProtocolError(Exception):
    """Raised when the engine returns a JSON-RPC error."""

    def __init__(self, code: int, message: str, data: ErrorData | None = None) -> None:
        super().__init__(message)
        self.code = code
        self.error_message = message
        self.data = data


def encode_request(request_id: int, method: str, params: dict[str, Any]) -> bytes:
    """Encode a JSON-RPC 2.0 request as NDJSON bytes.

    Returns compact JSON followed by newline, as bytes.
    """
    msg: dict[str, Any] = {
        "jsonrpc": "2.0",
        "id": request_id,
        "method": method,
        "params": params,
    }
    return json.dumps(msg, separators=(",", ":")).encode("utf-8") + b"\n"


def decode_response(line: bytes) -> dict[str, Any]:
    """Decode a JSON-RPC 2.0 response from NDJSON bytes.

    Returns the parsed response dict.
    Raises ValueError for malformed JSON.
    Raises ProtocolError if the response contains an error.
    """
    text = line.strip()
    if not text:
        raise ValueError("empty response line")

    try:
        data: Any = json.loads(text)
    except json.JSONDecodeError as e:
        raise ValueError(f"malformed JSON response: {e}") from e

    if not isinstance(data, dict):
        raise ValueError(f"expected JSON object, got {type(data).__name__}")

    if data.get("jsonrpc") != "2.0":
        raise ValueError(f"invalid jsonrpc version: {data.get('jsonrpc')}")

    if "error" in data and data["error"] is not None:
        err: dict[str, Any] = data["error"]
        error_data: ErrorData | None = None
        if "data" in err and err["data"] is not None:
            raw: dict[str, Any] = err["data"]
            error_data = ErrorData(
                error_type=raw.get("error_type", ""),
                retryable=raw.get("retryable", False),
                detail=raw.get("detail", ""),
            )
        raise ProtocolError(
            code=err.get("code", -1),
            message=err.get("message", "unknown error"),
            data=error_data,
        )

    return data


def extract_result(response: dict[str, Any]) -> Any:
    """Extract the result field from a decoded response."""
    if "result" not in response:
        raise ValueError("response missing 'result' field")
    return response["result"]


def extract_id(response: dict[str, Any]) -> int:
    """Extract the request ID from a decoded response."""
    if "id" not in response:
        raise ValueError("response missing 'id' field")
    return int(response["id"])

File: sdks/typescript/packages/core/src/proto/codec.ts
import type { ErrorData, RPCResponse } from "./types.js";
import { ProtocolError } from "./errors.js";

/**
 * Encode a JSON-RPC 2.0 request as an NDJSON line.
 */
export function encodeRequest(
  id: number,
  method: string,
  params: Record<string, unknown>,
): string {
  const msg = {
    jsonrpc: "2.0" as const,
    id,
    method,
    params,
  };
  return JSON.stringify(msg) + "\n";
}

/**
 * Decode a JSON-RPC 2.0 response from an NDJSON line.
 * Throws ProtocolError if the response contains an error.
 * Throws Error for malformed JSON or invalid responses.
 */
export function decodeResponse(line: string): RPCResponse {
  const text = line.trim();
  if (!text) {
    throw new Error("empty response line");
  }

  let data: unknown;
  try {
    data = JSON.parse(text);
  } catch (e) {
    throw new Error(`malformed JSON response: ${e instanceof Error ? e.message : String(e)}`);
  }

  if (typeof data !== "object" || data === null || Array.isArray(data)) {
    throw new Error(`expected JSON object, got ${Array.isArray(data) ? "array" : typeof data}`);
  }

  const obj = data as Record<string, unknown>;

  if (obj.jsonrpc !== "2.0") {
    throw new Error(`invalid jsonrpc version: ${String(obj.jsonrpc)}`);
  }

  if (obj.error != null) {
    const err = obj.error as Record<string, unknown>;
    let errorData: ErrorData | undefined;

    if (err.data != null) {
      const raw = err.data as Record<string, unknown>;
      errorData = {
        error_type: (raw.error_type as string) ?? "",
        retryable: (raw.retryable as boolean) ?? false,
        detail: (raw.detail as string) ?? "",
      };
    }

    throw new ProtocolError(
      (err.code as number) ?? -1,
      (err.message as string) ?? "unknown error",
      errorData,
    );
  }

  return obj as unknown as RPCResponse;
}

/**
 * Extract the result field from a decoded response.
 */
export function extractResult(response: RPCResponse): unknown {
  if (response.result === undefined) {
    throw new Error("response missing 'result' field");
  }
  return response.result;
}

/**
 * Extract the request ID from a decoded response.
 */
export function extractId(response: RPCResponse): number {
  if (response.id === undefined) {
    throw new Error("response missing 'id' field");
  }
  return Number(response.id);
}
